### ISSUE TITLE
Make API C++-friendly

### ISSUE THREAD

**Issue by legends2k** [recp/cglm#182]
In #83, I see one of the comments

> Also I would recommend against caring about or supporting C++. Realisitically no one would use this library over GLM if you're using C++. GLM is just a more mature, stable library. Keep your focus and energy on everything C related.

No, this isn't true.  I'm writing a C++ program and I like cglm _more than glm_ :)

I'd really appreciate if cglm has an API that' easy to use with a C++ compiler.  For example

``` c++
glm_lookat(vec3{50.f, 10.f, 20.f}, GLM_VEC3_ZERO, GLM_ZUP);
```

doesn't compile on g++ 10.  It screams

> Error: taking address of temporary

This is because I used the array API and so `vec3 eye` decays into `float *eye` and `vec3{50.f, 10.f, 20.f}` is a temporary.

Now I know **c**glm is a C-centric library and I appreciate and respect it.  However, [the idea proposed in #83 is pretty good][1]:

``` C
/* C and C++ version */
glm_vec3_norm((vec3){1, 2, 3});

/* C++ version if const X& is enabled otherwise: compiler error like 'No matching function for call to...' */
glm_vec3_norm({1, 2, 3});
```
> I think GLM_IN, GLM_OUT, GLM_INOUT macros will make C and C++ happy

This would be useful for the ones who prefer the array API over the struct one.  Thanks!

[1]: https://github.com/recp/cglm/issues/83#issuecomment-485280210

**Comment by recp** [comment]
@legends2k I agree with you. I think **cglm** could work better with C++. Added to my TODOs. 

With **GLM_IN**, **GLM_OUT**, **GLM_INOUT** macros, we could add compiler specific attributes maybe if we neeed any in the future. Let's think about it and try to find better solution. 

Also there is an issue with `const`, see https://github.com/recp/cglm/issues/83 maybe we could ignore `const` for C for now, I'll try to re-focus this issue later. 

Also, any feedbacks, help are always welcome

**Comment by awsdert** [comment]
Taking a cursory glance at this I'd say the simplest solution would be to have macros like glm_lookat (at least I assume it's a macro, haven't looked at details yet) have solid functions as alternatives so for example:
```var cglm_func_lookat( float *vec, uint pos, uint flags ) { return cglm_lookat( vec, pos, flags ); }```

**Comment by recp** [comment]
@awsdert thanks for your feedbacks,

Actually there is an interest to convert implemenntations to macro to avoid duplicate raw implementations for double:

https://github.com/recp/cglm/issues/196

so there is no need to prefix functions with `cglm_func_`. For C++, we could go with GLM_IN, GLM_OUT, GLM_INOUT which mentioned above

**Comment by PQCraft** [comment]
https://github.com/g-truc/glm

**Comment by legends2k** [comment]
@PQCraft GLM isn't that good with SIMD or inlining IIRC.  Of course, those may not be important to some projects.

**Comment by recp** [comment]
+1 for "Make API C++-friendly"

@legends2k thanks for sharing that, FWIW **cglm** implemen[ted]/[ing] **NEON** and other SIMDs or hardware features time by time so the comparison may not reflect latest cglm

**Comment by TheRektafire** [comment]
Personally I'm totally OK with cglm being made more c++ friendly as long as no performance is lost from having to stoop down to c++s level ;) if course if that becomes and issue I'm sure c and c++ versions of the functions could be offered separately

**Comment by awsdert** [comment]
I think it might be better to just use inline functions,  no macros, and revert to normal functions in C89 with some macros called BUILD_CGLM_ABI and BUILD_CGLM_LIB, the lib would be for *.so/*.dll libraries while the abi would be for defining the contents of the functions that otherwise would be just declarations. If inlining is available then the abi macro can be ignored providing the speedup that later specifications allow but gracefully falling back to normal functions when inlining is not available, in both cases proper typedefs can be used, C++ is perfectly happy to use both so it's not a problem, at most you would slap some wrapper classes around the inline/normal functions and maybe some overloaded functions for anything not suited for a class

**Comment by NotAFreak** [comment]
Sorry in advance for necroing a 5-year old thread.
I'm using cglm in the short term as glm is having a lot of compile issues for me, even though i prefer the style of glm's api, and I just want to get progress on my engine done. Moving my code to continue development from my desktop to my laptop, I've come across the same "error: taking address of temporary array", on Ubuntu 24.04 LTS. Fairly annoying. 
I would like to note that I don't come across these compile issues while using Clang, I do in GCC however.

**Comment by recp** [comment]
@NotAFreak Thanks for the feedback, i’ve been quite busy for a while (and still am 🙂), but I’ll review the issue and the suggestions asap so we can move forward, more feedback would helpful

**Comment by awsdert** [comment]
Just a note on the inline thing, you can avoid repeating yourself in the build files with something like this:

foo.h:
```C
...
#ifndef FOO_C
#define FOO_C
#define FOO_X 0
#else
#define FOO_X 1
#endif
#if STDC_VERSION >= 199100L
FOO_C inline
#include <foo.bar.h>
#include <foo.bar.c>
#else
FOO_API
#include <foo.bar.h>
#ifndef FOO_X
;
#else
#include <foo.bar.c>
#endif
#endif
```
foo.bar.h:
```C
int bar(void)
```
foo.bar.c:
```C
{ return 0; }
```
foo.c:
```C
#define FOO_C extern
#include <foo.h>
...
```
Not perfect by any means but it does help if the code is big

**Edit:** Alternatively:

foo.h:
```C
...
#ifndef FOO_C
#define FOO_C
#endif
#if STDC_VERSION >= 199100L
FOO_C inline
#include <foo.bar.h>
#include <foo.bar.c>
#else
FOO_API
#include <foo.bar.h>
;
#endif
```
foo.c:
```C
#define FOO_C extern
#include <foo.h>
#if STDC_VERSION < 199100L
FOO_API
#include <foo.bar.h>
#include <foo.bar.c>
#endif
...
```
Might be easier to read
